tr/tr.1
.\"
.\" Sccsid @(#)tr.1	1.21 (gritter) 8/6/05
.\" Parts taken from tr(1), Unix 7th edition:
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" SPDX-Licence-Identifier: Caldera
.\"
.TH TR 1 "8/6/05" "Heirloom Toolchest" "User Commands"
.SH NAME
tr \- translate characters
.SH SYNOPSIS
\fBtr\fR [\fB\-cCds\fR] [\fIstring1\fR [\fIstring2\fR]]
.SH DESCRIPTION
.I Tr
copies the standard input to the standard output with
substitution or deletion of selected characters.
Input characters found in
.I string1
are mapped into the corresponding characters of
.IR string2 .
When
.I string2
is shorter than
.IR string1 ,
no mapping occurs beyond its last character.
Any combination of the options
.B \-cds
may be used:
.TP
.B \-c
complements the set of characters in
.I string1
with respect to the universe of characters
whose byte codes are 01 through 0377 octal,
or with a multibyte character set,
whose wide character codes start at 1.
.TP
.B \-d
deletes all input characters in
.I string1.
.TP
.B \-s
squeezes all strings of repeated output characters that are
in
.I string2
to single characters.
.PP
The following option has been introduced by POSIX.1-2001:
.TP
.B \-C
complements the set of characters in
.I string1
like
.IR \-c ,
but orders ranges according to the collation sequence.
.PP
In either string the following character sequences are
treated specially:
.TP 8
\fB\e\fIoctal\fR
The character
`\e' followed by 1, 2 or 3 octal digits stands for the
character whose byte code is given by those digits.
Multibyte characters can be specified
as a sequence of octal bytes.
.TP
\fB\e\fIchar\fR
The escape sequences `\ea' (bell),
`\eb' (backspace),
`\ef' (form feed),
`\en' (newline),
`\er' (carriage return),
`\et' (horizontal tabulator),
and `\ev' (vertical tabulator)
are supported.
A `\e' followed by any other character (other than an octal digit)
stands for that character.
.TP 8
\fB[\fIa\fB\-\fIz\fB]\fR
(/usr/5bin/tr, /usr/5bin/s42/tr)
means a range of characters from
.I a
to
.I z
in increasing byte order,
or with a multibyte character set,
in increasing wide character order.
With the
.I \-C
option,
characters are ordered according to the collation sequence.
.TP 8
\fIa\fB\-\fIz\fR
(/usr/5bin/posix/tr, /usr/5bin/posix2001/tr)
means a range of characters from
.I a
to
.I z
in increasing byte order,
or with a multibyte character set,
in increasing wide character order.
With the
.I \-C
option,
characters are ordered according to the collation sequence.
.TP 8
\fB[:\fIclass\fB:]\fR
means all characters that belong to character class
.I class
in the current LC_CTYPE locale
in increasing byte order,
or with a multibyte character set,
in increasing wide character order.
With the
.I \-C
option,
characters are ordered according to the collation sequence.
If both \fB[:upper:]\fR and \fB[:lower:]\fR appear
at the same position in either string,
upper-case characters are mapped to lower-case characters
(and vice versa).
.TP 8
\fB[=\fIc\fB=]\fR
where \fIc\fR is a collating symbol in the current LC_COLLATE locale,
means all characters that belong to the same equivalence class as \fIc\fR,
i.\ e. have the same collating weight as \fIc\fR.
.TP 8
\fB[\fIa\fB*\fIn\fB]\fR
means
.I n
repetitions
of the character
.IR a ,
with
.I n
as an octal number if it starts with `0'
and as a decimal number otherwise.
If
.I n
is omitted or zero,
it is taken to be huge
(useful for padding
.I string2
to the length of
.IR string1 ).
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale (7).
.TP
.B LC_COLLATE
Affects the composition of equivalence classes.
.TP
.B LC_CTYPE
Determines the mapping of bytes to characters
in translation strings and input files, and
the availability and composition of character classes.
.SH EXAMPLES
The following examples create a list of all
the words in `file1' one per line in `file2',
where a word is taken to be a maximal string of alphabetics.
The strings are quoted
to protect `\e' and '[' from the Shell.
012 is the ASCII code for newline.
.IP \& 2
/usr/5bin/tr \-cs \'[A\-Z][a\-z]\' \'[\e012*]\' <file1 >file2
.IP \& 2
/usr/5bin/s42/tr \-cs \'[A\-Z][a\-z]\' \'[\e012*]\' <file1 >file2
.IP \& 2
/usr/5bin/posix/tr \-cs A\-Za\-z \'[\e012*]\' <file1 >file2
.IP \& 2
/usr/5bin/posix2001/tr \-cs A\-Za\-z \'[\e012*]\' <file1 >file2
.SH "SEE ALSO"
ed(1),
ascii(7),
locale(7)
.SH NOTES
/usr/5bin/tr and /usr/5bin/s42/tr do not handle ASCII NUL in
.I string1
or
.IR string2 ;
they always delete NUL from input.
.PP
The LC_COLLATE variable is not respected;
equivalence classes consist of exactly one character,
and the
.I \-c
and
.I \-C
options produce identical results.
.PP
Portable programs must prefix the
.RB ` [ '
and
.RB ` \- '
characters with a backslash and cannot use them otherwise,
unless a `\fB[\fIa\fB\-\fIz\fB]\fR' sequence appears at
exactly the same position in both operand strings.
The `\fB\e\fIoctal\fR' construct
cannot be used portably preceding or following the `\fB\-\fR'
range metacharacter;
the `\fB\e\fIchar\fR' escapes for control characters
are not present in old implementations,
use octal escapes instead.


CHANGES
Release 250220
* seq, write: Fix memory fault due to illegal pointer access.
* cpio, pax: Fix manual page formatting.

Release 250209
* cpio: Fix manual page formatting.
* apply: Fixed the command string magic character computation algorithm that
  got garbled after a refactor made in the name of performance.
* build/mk.config, chroot, nawk: Adding -ltinfo for static compilation with
  glibc and making changes to adhere to C standards (patch by Gabriel Souza).
* [NEW] wall, write: New implementation ported from UNIX 7th (a.k.a. 32V),
  bulletproofed against buffer overflows and the CVE-2024-28085, which is
  available as the 'ssafe()' function at libcommon.
* rm: Alias '-e' flag to more known '-v' for verbose.
* date, shl: Fix compilation on platforms where (struct utmpx).ut_tv is not
  of type 'timeval'.
* rev: UTF-8 support, borrowed from suckless sbase. In varietate concordia.
* [NEW] apply, pick: Complete implementation for both apply and pick, from
  Research UNIX 8th and 10th respectively; extended payloads for pick in
  comparison with the original implementation.
* seq: Support for counting from negative numbers to positive or making a
  negative sequence.
* rev: Full refactor (made by Samuel Brederodes).
* Fixes for building dynamically on Arch Linux and other GNU C
  Library-based Linux systems which ships exclusively dynamic versions
  of its libraries, also remove `readlink_ucb' object file at `readlink/'
  build directory when executing `make clean' (patch by Molly A. McCollum).
* Remove `register's from legacy code that was living unchanged since at
  least the 80's, since modern linters started complaining about it and
  it is not used as much on most common architectures such as x86
  (patch by Samuel Brederodes).
* watch: `use_default_colors()', do not force black and white.
* [NEW] rev: New implementation, not capable of UTF-8 yet.
* cpio, diff, tabs: Fix duplicated definitions (made/reported by crumbtoo,
  merged by me, Luiz Antônio Rangel). Thanks! --- or better: Merci!
* rm: Added real support for the POSIX 1003.1(2008) standard, implemented
  IBM AIX(R)-style verbosing and 4.4BSD's '-d' option for directories ---
  rmdir(1)-esque payload --- as extensions to the standard.
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants.

Release 230503
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin.

Release 230120
* Notes for the library directories if linking against the GNU C library.

tr/tr.c
/*
 * Derived from Unix 6th Edition /usr/source/s2/tr.c and
 * Unix 7th Edition /usr/src/cmd/tr.c
 *
 * Changes by Gunnar Ritter, Freiburg i. Br., Germany, August 2005.
 */
/*
 * Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
 *
 * SPDX-Licence-Identifier: Caldera
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
#if defined (SUS)
static const char sccsid[] USED = "@(#)tr_sus.sl	2.3 (gritter) 9/7/05";
#elif defined (UCB)
static const char sccsid[] USED = "@(#)/usr/ucb/tr.sl	2.3 (gritter) 9/7/05";
#else
static const char sccsid[] USED = "@(#)tr.sl	2.3 (gritter) 9/7/05";
#endif

#include <stdlib.h>
#include <stdio.h>
#include <limits.h>
#include <locale.h>
#include <unistd.h>
#include <wchar.h>
#include <wctype.h>
#include <string.h>
#include <inttypes.h>
#include "iblok.h"

#if defined (__GLIBC__) && defined (_IO_putc_unlocked)
#undef	putchar
#define	putchar(c)	_IO_putc_unlocked(c, stdout)
#endif

#undef	iswupper
#undef	iswlower
#undef	towupper
#undef	towlower

#ifdef	SUS
#define	NIL	EOF
#else
#define	NIL	0
#endif

/* tr - transliterate data stream */

/*
 * This implementation of tr keeps the traditional design to hold
 * arrays that contain all characters. This is because experience
 * with tr implementations that use more sophisticated structures
 * has shown that their memory usage is even higher under certain
 * circumstances while their performance is generally worse. This
 * is mostly because these structures usually contain one or more
 * pointers in addition to the actual data for a character; these
 * pointers consume memory and need to be followed. The backdraw
 * with this implementation is that it is expected that the
 * contents of wchar_t values are all within the Unicode range.
 *
 * Tables of different sizes are built as needed. Thus as long as
 * only ASCII characters occur, only a few kilobytes of memory are
 * consumed in any locale. If all characters in a multibyte locale
 * remain within the BMP, around 250 kilobytes are needed. The full
 * Unicode range causes tr to use about 4.5 megabytes of memory.
 * The -c or -s options raise these requirements by up to 125 %
 * to a total maximum of about 10 megabytes.
 */

static int dflag = 0;
static int sflag = 0;
static int cflag = 0;
static int save = NIL;
static int32_t *code;
static char *squeez;
static int32_t *vect;
static struct string {
	int last, max, rep;
	wctype_t type;
	enum { N, U, L } tp;
	char *p;
} string1, string2;
static int status;
static int ccnt;
static int xccnt;
static int mask;
static int mb_cur_max;

static void cmple(char *, char *, int);
static int next(struct string *);
static int nextc(struct string *);
static int nextb(struct string *);
static void bad(void);
static int caseconv(wint_t (*)(wint_t), int (*)(wint_t));

int
main(int argc, char **argv)
{
	int i, j;
	int c;
	char *arg0, *arg1;
	struct iblok *ip;

	setlocale(LC_CTYPE, "");
	mb_cur_max = MB_CUR_MAX;
	xccnt = mb_cur_max > 1 ? 0x110000 : 256;
	mask = mb_cur_max > 1 ? 0x1FFFFF : 255;

	ip = ib_alloc(0, 0);

	while((c = getopt(argc, argv, ":cCds")) != EOF) {
		switch(c) {
		case 'c':
		case 'C':
			cflag++;
			break;
		case 'd':
			dflag++;
			break;
		case 's':
			sflag++;
			break;
		}
	}
	argc -= optind;
	argv += optind;
	arg0 = argc>0 ? argv[0] : "";
	arg1 = argc>1 ? argv[1] : "";

	cmple(arg0, arg1, 255);

	if(mb_cur_max > 1) {
		char mb[MB_LEN_MAX];
		wint_t wc;
		char *cp;
		while((cp=ib_getw(ip, &wc, &i)) != NULL) {
			if(wc == NIL) continue;
			if(wc!=WEOF && (wc&mask)>=ccnt)
				cmple(arg0, arg1, wc);
			if(wc!=WEOF && (wc&=mask)<ccnt) {
				if((c = code[wc]) != NIL) {
					if(!sflag || c!=save || !squeez[c]) {
						save = c;
						if(c != wc) {
							if (c & ~0177) {
								j=wctomb(mb, c);
								if(j>0) {
									cp = mb;
									i = j;
								}
							} else {
								i = 1;
								cp = mb;
								*cp = c;
							}
						}
					} else
						continue;
				} else
					continue;
			}
			for(j = 0; j < i; j++)
				putchar(cp[j]&0377);
		}
	} else {
		while((c=ib_get(ip)) != EOF) {
			if(c == NIL) continue;
			if((c = code[c]) != NIL)
				if(!sflag || c!=save || !squeez[c])
					putchar(save = c);
		}
	}
	return status;
}

static void
cmple(char *arg0, char *arg1, int needc)
{
	int i, j;
	int c, d;
	int32_t *compl = NULL;
	int lastd = NIL;
	int nccnt = 0;
	int32_t *ncode, *nvect = NULL;
	char *nsqueez = NULL;
	const int borders[] = {
		256,		/* singlebyte */
		0x3000,		/* Western and technical Unicode characters */
		0x10000,	/* whole BMP */
		0x30000,	/* most of Unicode 4.0 */
		0x110000,	/* whole Unicode range */
		-1
	};

	for(i=0; borders[i]>0; i++)
		if((nccnt=borders[i]) > needc) break;
	if(nccnt<=ccnt) return;

	ncode = realloc(code, nccnt * sizeof *code);
	if(sflag) nsqueez = realloc(squeez, nccnt * sizeof *squeez);
	if(cflag) nvect = realloc(vect, nccnt * sizeof *vect);
	if(ncode==NULL || (sflag && nsqueez==NULL) || (cflag && nvect==NULL))
		return;
	ccnt = nccnt;
	code = ncode;
	squeez = nsqueez;
	vect = nvect;
	if(code==NULL || (sflag && squeez==NULL) || (cflag && vect==NULL)) {
		write(2, "No memory\n", 10);
		_exit(077);
	}
	if(sflag) memset(squeez, 0, nccnt * sizeof *squeez);

	string1.p = arg0;
	string2.p = arg1;
	string1.last = string2.last = NIL;
	string1.max = string2.max = NIL;
	string1.rep = string2.rep = 0;
	string1.tp = string2.tp = N;
	string1.type = string2.type = 0;

	for(i=0; i<ccnt; i++)
		code[i] = NIL;
	if(cflag) {
		for(i=0; i<ccnt; i++)
			vect[i] = NIL;
		while((c = next(&string1)) != NIL)
			vect[c] = 1;
		j = 0;
		for(i=NIL+1; i<ccnt; i++)
			if(vect[i]==NIL) vect[j++] = i;
		vect[j] = NIL;
		compl = vect;
	}
	for(;;){
		if(cflag) c = *compl++;
		else c = next(&string1);
		if(c==NIL) break;
		d = next(&string2);
		if(string1.tp==U && string2.tp==L) {
			lastd = caseconv(towlower, iswlower);
			continue;
		}
		if(string1.tp==L && string2.tp==U) {
			lastd = caseconv(towupper, iswupper);
			continue;
		}
		string1.tp = string2.tp = N;
#ifndef	UCB
		if(d==NIL) d = c;
#else
		if(d==NIL) d = lastd;
#endif
		lastd = d;
		if(c<ccnt) code[c] = d;
		if(d<ccnt && sflag) squeez[d] = 1;
	}
	free(vect);
	while((d = next(&string2)) != NIL) {
		if(sflag) squeez[d] = 1;
		if(string2.max==NIL && (string2.p==NULL || *string2.p==0))
			break;
	}
	for(i=0;i<ccnt;i++) {
		if(code[i]==NIL) code[i] = i;
		else if(dflag) code[i] = NIL;
	}
}

static int
next(struct string *s)
{
	int a, c, n;
	int base;
	char *savep;

	if(--s->rep > 0) return(s->last);
	if(s->max != NIL) {
		if(s->last++ < s->max) {
			if(s->type) {
				do {
					c = mb_cur_max==1 ?
						btowc(s->last) : s->last;
					if(iswctype(c, s->type))
						break;
				} while(s->last++ < s->max);
				if(s->last > s->max) goto skip;
			}
			return(s->last);
		}
	skip:	s->max = s->last = NIL;
		s->tp = N;
		s->type = 0;
	}
#if defined (SUS) || defined (UCB)
	if(s->p && s->last != NIL && *s->p=='-') {
		nextc(s);
		s->max = nextc(s);
		if(s->max==NIL)
			return('-');
		if(s->max < s->last)  {
			s->last = s->max-1;
			return('-');
		}
		return(next(s));
	}
#endif	/* SUS || UCB */
	savep = s->p;
	if(s->p && *s->p=='[') {
		nextc(s);
		s->last = a = nextc(s);
		if(savep[1]==':') {
			char	*sp;
			for(sp=s->p; *sp && *sp != ':'; sp++);
			if(sp[0]==0 || sp[1]!=']') bad();
			*sp=0;
			if((s->type = wctype(s->p)) == 0) bad();
			if(strcmp(s->p,"upper")==0) s->tp = U;
			else if(strcmp(s->p,"lower")==0) s->tp = L;
			*sp=':';
			s->max = ccnt-1;
			s->last = NIL;
			s->p = &sp[2];
			return(next(s));
		}
		if(a=='=') {
			s->last = nextc(s);
			if(s->p[0]!='=' || s->p[1]!=']') bad();
			nextc(s); nextc(s);
			return s->last;
		}
		s->max = NIL;
		switch(s->p ? *s->p : 0) {
#if defined (SUS) || defined (UCB)
		default:
			s->p = savep;
			break;
#else	/* !SUS, !UCB */
		case '-': {
			int b;
			nextc(s);
			b = nextc(s);
			if(b<a || *s->p++!=']')
				bad();
			s->max = b;
			return(a);
		    }
		default:
			bad();
#endif	/* !SUS, !UCB */
		case '*':
			nextc(s);
			base = (*s->p=='0')?8:10;
			n = 0;
			while((c = *s->p)>='0' && c<'0'+base) {
				n = base*n + c - '0';
				s->p++;
			}
			if(*s->p++!=']') bad();
			if(n==0) n = INT_MAX;
			s->rep = n;
			return(a);
		}
	}
	return(s->last = nextc(s));
}

static int
nextc(struct string *s)
{
	char mb[MB_LEN_MAX+1];
	wchar_t wc;
	int c, i, n;
	mbstate_t state;

	if(s->p == NULL)
		return NIL;
	c = nextb(s);
	if(c!=NIL && mb_cur_max>1 && c&0200) {
		i = 0;
		for(;;) {
			mb[i++] = c;
			mb[i] = '\0';
			memset(&state, 0, sizeof state);
			if((n = mbrtowc(&wc, mb, i, &state))==(size_t)-1)
				bad();
			if(n==(size_t)-2) {
				if((c = nextb(s))==NIL) bad();
				continue;
			}
			return(wc);
		}
	}
	return(c);
}

static int
nextb(struct string *s)
{
	int c, i, n = -1;

	if (s->p == NULL)
		return NIL;
	c = *s->p++;
	if(c=='\\') {
		i = n = 0;
		while(i<3 && (c = *s->p)>='0' && c<='7') {
			n = n*8 + c - '0';
			i++;
			s->p++;
		}
		if(i>0) c = n;
		else {
			switch (c = *s->p++) {
			case 'a': c = '\a'; break;
			case 'b': c = '\b'; break;
			case 'f': c = '\f'; break;
			case 'n': c = '\n'; break;
			case 'r': c = '\r'; break;
			case 't': c = '\t'; break;
			case 'v': c = '\v'; break;
			}
		}
	}
	if (s->p[-1] == '\0')
		s->p = NULL;
	return(c || n>=0 ? c&0377 : NIL);
}

static void
bad(void)
{
	write(2,"Bad string\n",11);
	_exit(2);
}

static int
caseconv(wint_t (*to)(wint_t), int (*is)(wint_t))
{
	int c, d = NIL, i;

	string1.tp = string2.tp = N;
	string1.type = string2.type = 0;
	string1.last = string2.last = NIL;
	string1.max = string2.max = NIL;
	for(i=0;i<ccnt;i++) {
		c = mb_cur_max==1 ? btowc(i) : i;
		if(c!=WEOF) {
			d = (*to)(c);
			code[i] = mb_cur_max==1 ? wctob(d) : d;
			if(sflag && d<ccnt && (*is)(d)) squeez[d] = 1;
		}
	}
	return(d);
}
